packages/mui-material/src/MenuList/MenuList.js

```diff
@@ -235,7 +235,10 @@ const MenuList = React.forwardRef(function MenuList(props, ref) {
 
     if (!child.props.disabled) {
       if (variant === 'selectedMenu' && child.props.selected) {
+        lastSelectedActiveItemIndex.current = index;
         activeItemIndex = index;
+      } else if (lastSelectedActiveItemIndex.current !== -1) {
+        activeItemIndex = lastSelectedActiveItemIndex.current;
       } else if (activeItemIndex === -1) {
         activeItemIndex = index;
       }
```

Comment:
I think this is not solving the issue 🤔

The rule we want to add here is "if a children changed, then focus that one, otherwise, continue with the current behavior". The following code achieves this:

```js
// outside the children forEach
const prevChildren = usePreviousProps(children);
let childSelectedChanged = false;

// inside the children forEach, on line 236
if (!child.props.disabled) {
  const isSelected = child.props.selected;
  const wasSelected = prevChildren[index]?.props.selected;
  if (variant === 'selectedMenu') {
    if (wasSelected !== isSelected) {
      childSelectedChanged = true;
      activeItemIndex = index;
    } else if (!childSelectedChanged && isSelected) {
      activeItemIndex = index;
    }
  } else if (activeItemIndex === -1) {
    activeItemIndex = index;
  }
}
```

I think that should do it, but please test it to see if it covers all the cases the PR is trying to solve. If we go with this solution, we should update the description of the algorithm which is above the children `forEach`.

Note: This assumes only one child can change at each render, which I think is correct

Reply:
It's working only one way around. It works when the deselected item is above, but not when it's below:

https://github.com/mui/material-ui/assets/16889233/e748d951-a256-410c-9f29-9ca064d81a78